google-cloud-pubsub/lib/google/cloud/pubsub/subscriber.rb

```diff
@@ -169,6 +170,87 @@ def stopped?
           synchronize { @stopped }
         end
 
+        ##
+        # Register to be notified of errors when raised.
+        #
+        # If an unhandled error has occurred the subscriber will attempt to
+        # recover from the error and resume listening.
+        #
+        # @yield [callback] The block to be called when an error is raised.
+        # @yieldparam [Exception] error The error raised.
+        #
+        # @example
+        #   require "google/cloud/pubsub"
+        #
+        #   pubsub = Google::Cloud::Pubsub.new
+        #
+        #   sub = pubsub.subscription "my-topic-sub"
+        #
+        #   subscriber = sub.listen do |received_message|
+        #     # process message
+        #     received_message.acknowledge!
+        #   end
+        #
+        #   # Register to be notified when unhandled errors occur.
+        #   subscriber.on_error do |error|
+        #     # log error
+        #     puts error
+        #   end
+        #
+        #   # Start listening for messages and errors.
+        #   subscriber.start
+        #
+        #   # Shut down the subscriber when ready to stop receiving messages.
+        #   subscriber.stop.wait!
+        #
+        def on_error &block
+          synchronize do
+            @error_callback = block
+          end
+        end
+
+        ##
+        # The most recent unhandled error to occur while listening to messages
+        # on the subscriber.
+        #
+        # If an unhandled error has occurred the subscriber will attempt to
+        # recover from the error and resume listening.
+        #
+        # @return [Exception, nil] error The most recent error raised.
+        #
+        # @example
+        #   require "google/cloud/pubsub"
+        #
+        #   pubsub = Google::Cloud::Pubsub.new
+        #
+        #   sub = pubsub.subscription "my-topic-sub"
+        #
+        #   subscriber = sub.listen do |received_message|
+        #     # process message
+        #     received_message.acknowledge!
+        #   end
+        #
+        #   # Start listening for messages and errors.
+        #   subscriber.start
+        #
+        #   # If an error was raised, it can be retrieved here:
+        #   subscriber.last_error #=> nil
+        #
+        #   # Shut down the subscriber when ready to stop receiving messages.
+        #   subscriber.stop.wait!
+        #
+        def last_error
+          synchronize { @last_error }
+        end
+
+        # @private returns error object from the stream thread.
+        def error! error
+          synchronize do
+            @last_error = error
+            @error_callback.call error unless @error_callback.nil?
```

Comment:
Are you sure you need the synchronization in these methods? (Remember the lock is being shared with the start-stop synchronization elsewhere in this file.) It looks like the only race we're dealing with here is the modification and usage of `@error_callback` which can be handled without explicit locks by replacing this line with

```ruby
error_callback = @error_callback
error_callback.call error unless error_callback
```

Reply:
It is possible that multiple Stream objects will want to make calls on the parent Subscriber object concurrently. My intention here was that setting `@last_error` and calling the error callback would happen together. So if for some reason the user-provided error callback called `#last_error` it would get the same object.

Reply:
I see your point though, we don't want to actually lock during the callback. I'll change this.

Reply:
Updated this to not call the callbacks in the `synchronize ` block.